src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class Instrumentation
+    {
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (methodBody != null)
+            {
+                MethodSymbol createPayload = GetCreatePayload(compilation, diagnostics);
+                MethodSymbol flushPayload = GetFlushPayload(compilation, diagnostics);
+
+                // Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it is instrumented.
+                if ((object)createPayload != null && (object)flushPayload != null && (object)method != createPayload && (object)method != flushPayload)
+                {
+                    // Create the symbol for the instrumentation payload.
+                    SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
+                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
+                    TypeSymbol payloadElementType = boolType;
+                    ArrayTypeSymbol payloadType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, payloadElementType);
+                    FieldSymbol payloadField = GetPayloadField(method, methodOrdinal, payloadType, factory, compilation);
+
+                    // Synthesize the instrumentation and collect the spans of interest.
+
+                    BoundBlock newMethodBody;
+                    InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, payloadField, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans, out newMethodBody);
+
+                    // Synthesize the initialization of the instrumentation payload array. It should be done either statically or with concurrency-safe code:
+                    //
```

Comment:
I see. So we can guarantee both of those constraints?
1. payloadField is atomically null or is ready, but nothing in between
2. Multiple calls to Instrumentation.CreatePayload don't cause harm